lexos/models/rolling_windows_model.py

```diff
@@ -0,0 +1,574 @@
+    RollingWindowsReceiver, WindowUnitType, RWATokenType
+
+# Set the rwa regex flags.
+rwa_regex_flags = re.DOTALL | re.MULTILINE | re.UNICODE
+
+
+class RWATestOptions(NamedTuple):
+    """RWA test front end options."""
+
+    file_id_content_map: FileIDContentMap
+    rolling_windows_options: RWAFrontEndOptions
+
+
+class RollingWindowsModel(BaseModel):
+    """The class for rolling window calculation."""
+
+    def __init__(self, test_option: Optional[RWATestOptions] = None):
+        """Initialize the class based on if test option was passed in.
+
+        :param test_option: the options to send in for testing.
+        """
+        super().__init__()
+        if test_option is not None:
+            self._test_file_id_content_map = test_option.file_id_content_map
+            self._test_front_end_options = test_option.rolling_windows_options
+        else:
+            self._test_file_id_content_map = None
+            self._test_front_end_options = None
+
+    @property
+    def _passage(self) -> str:
+        """Get the passage to run rolling windows on.
+
+        :return: the content of the passage as a string.
+        """
+        # if test option is specified
+        if self._test_file_id_content_map is not None and \
+                self._test_front_end_options is not None:
+            file_id = self._test_front_end_options.passage_file_id
+            file_id_content_map = self._test_file_id_content_map
+
+        # if test option is not specified, get option from front end
+        else:
+            file_id = RollingWindowsReceiver().options_from_front_end() \
+                .passage_file_id
+            file_id_content_map = FileManagerModel().load_file_manager() \
+                .get_content_of_active_with_id()
+
+        return file_id_content_map[file_id]
+
+    @property
+    def _options(self) -> RWAFrontEndOptions:
+        """Get the front end option packed as one named tuple.
+
+        :return: a RWAFrontEndOption packs all the frontend option.
+        """
+        return self._test_front_end_options \
+            if self._test_front_end_options is not None \
+            else RollingWindowsReceiver().options_from_front_end()
+
+    @staticmethod
+    def _get_rolling_window_from_list(input_list: List[str],
+                                      window_size: int) -> np.ndarray:
+        """Get the rolling window from the list of terms.
+
+        :param input_list: A list of terms (word, char or line),
+            depends on the window type (word and line are with endings).
+        :param window_size: The size of the window (number of terms in window).
+        :return: An array of strings, each element is a window.
+        """
+        # Number of items in the input list.
+        num_item = len(input_list)
+
+        # Get the total number of windows.
+        num_window = num_item - window_size + 1
+
+        # Get the rolling list, should be a array of str.
+        return np.array([
```

Comment:
@chantisnake After some testing around, I found this line is extremely slow when it comes to the case that user specifies a large window size. Assume we have a file that is 500KB (which isn't large at all, apparently we support file size up to 250MB) and we specified the window size to be 10000, I think we will end up saving something around `500KB * 10000`. And in fact, this often causes Lexos to crash with exit code `247` when windows size is big. I didn't find much details about this exiting code but I assume it is memory related. I have another branch that I tried to get rid of this line, so we don't need to save all windows. It won't crash at that point but it still take a really long time to process. Any suggestions on improving this will be appreciated.

Reply:
Hum, I am guessing this is caused by the join function being slow. don't really know how to fix this.

Reply:
I have experimented with  `join` method they are super fast. 
```python
In [22]: import uuid  # use to generate random string

In [23]: a = [str(uuid.uuid4()) for _ in range(10000)] # 10000 random strings

In [24]: a[:5]  # these are long random strings
Out[24]:
['752779e9-3260-46c7-9072-c3604d45711c',
 '8e5526e8-74bb-4fc6-83dc-33dc8f709fb0',
 '4484acff-758d-4b9d-a72f-eb45e1c59e31',
 '143a168e-4df5-421d-a8b9-db03b1b969f8',
 'dfdec4f2-a0ce-45eb-ab2d-10691ac82a0d']

In [25]: %time b = "".join(a)  # time the concatenation
Wall time: 0 ns

In [26]: b[0:500]  # to preview that they are really joined
Out[26]: '752779e9-3260-46c7-9072-c3604d45711c8e5526e8-74bb-4fc6-83dc-33dc8f709fb04484acff-758d-4b9d-a72f-eb45e1c59e31143a168e-4df5-421d-a8b9-db03b1b969f8dfdec4f2-a0ce-45eb-ab2d-10691ac82a0d5951c65c-6fc8-4329-b79a-6c0007406fb00feb4e73-251b-4a55-8455-1ebed1d2834fa627f844-0f00-4d5f-9725-5342f6dda46d58064618-789d-42ae-bccb-a181796bc95152c774ad-572a-47c6-b380-cf9482c89ba0ca2b91b6-6032-4999-9f98-90e8a1c34c4934450420-a4c6-4388-878f-d32992d593b91dc18257-0971-488d-9bb6-0d23df80daa5577f0422-ec1c-4e7c-88d9-cfe6cb4a'
```

There are no way they can cause lag. I don't know the reason for this.